keps/sig-instrumentation/3466-healthcheck-metrics/README.md

```diff
@@ -0,0 +1,315 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP intends to allow us to emit health check data in a structured and consistent way,
+so that monitoring agents can consume healthz/livez/readyz data and create SLOs/alerts off
+of these inputs.
+
+## Motivation
+
+Healthchecking data is currently surfaced in unstructured format and is scraped by monitoring
+agents (as well as the kubelet), which must be configured to interpret the health data and
+to act upon it. This process does not lend itself readily to the creation of availability SLOs,
+since we basically require an outside agent to parse the health endpoint and convert this signal
+into an SLI. 
+
+### Goals
+
+- Create a uniform interface by which we can consume health checking information
+- Allow availability SLIs to be created without a specialized monitoring agent
+- Allow for increased granularity (by configuring a smaller scrape interval) of health check metric data
+- Minimize the diff involved for each Kubernetes component
+
+
+### Non-Goals
+
+- Creation of SLOs are out of scope. This KEP specifically targets the creation of signals which can be used as SLIs.
+
+## Proposal
+
+We are proposing to add a new endpoint on Kubernetes components `/metrics/health` which returns
+health check data in prometheus format.
+
+
+### Risks and Mitigations
+
+This is a separate endpoint, it does not have to be used. The risk of adding metrics is generally cardinality, but in this
+case we are proposing known dimensions to the metrics, specifically:
+
+1. status - one of `Success`, `Error`, `Pending`
+1. type - one of `livez`, `readyz`, `healthz`
+1. name - the known name of the health check. AFAIK, these are all static strings in the Kubernetes codebase, therefore bounded in cardinality
+
+
+## Design Details
+
+We will write the actual metric data in the healthz/livez/readyz paths and expose this in our new
```

Comment:
I think you're trying to say that it will call the existing healthz/livez/readyz code paths (or hit the pages?), convert that data into metric form, and serve it?

Reply:
Or are you trying to say that when those pages are called, you'll cache the result?

Reply:
Ah, I think you are trying to say, "When healthz/livez/readyz paths are accessed (not on a timer), they will record whatever they return in a gauge metric."

Reply:
Yeah that's what I was trying to say. Lemme rephrase it.